src/stats/histograms.jl

```diff
@@ -71,28 +72,75 @@ function Base.quantile(o::Hist, p = [0, .25, .5, .75, 1])
     quantile(mids[inds], fweights(counts[inds]), p)
 end
 
-#-----------------------------------------------------------------------# FixedRangeBins
-mutable struct FixedRangeBins{R <: Range} <: HistAlg{0} 
-    edges::R 
+#-----------------------------------------------------------------------# FixedBins
+mutable struct FixedBins{closed,E<:AbstractVector} <: HistAlg{0}
+    edges::E
     counts::Vector{Int}
     out::Int
+
+    function FixedBins{closed,E}(edges::E, counts::Vector{Int},
+                                 out::Int) where {E<:AbstractVector,closed}
+        closed == :left || closed == :right ||
+            error("closed must be left or right")
+        length(edges) == length(counts) + 1 ||
+            error("Histogram edge vectors must be 1 longer than corresponding count vectors")
+        issorted(edges) || error("Histogram edge vectors must be sorted in ascending order")
+        new{closed,E}(edges, counts, out)
+    end
 end
-get_hist_alg(r::Range) = FixedRangeBins(r, zeros(Int, length(r) - 1), 0)
-_midpoints(o::FixedRangeBins) = _midpoints(o.edges)
-_counts(o::FixedRangeBins) = o.counts
-nobs(o::FixedRangeBins) = sum(o.counts) + o.out
-function fit!(o::FixedRangeBins, y, γ::Number)
-    r = o.edges 
-    a = first(r)
-    δ = step(r)
-    k = floor(Int, (y-a) / δ) + 1
-    if 1 ≤ k < length(r)
-        @inbounds o.counts[k] += 1
-    else 
+Base.@pure FixedBins(edges::AbstractVector, counts::Vector{Int}, out::Int; closed::Symbol = :left) =
+    FixedBins{closed,typeof(edges)}(edges, counts, out)
+
+get_hist_alg(e::AbstractVector; kwargs...) = FixedBins(e, zeros(Int, length(e) - 1), 0; kwargs...)
+_midpoints(o::FixedBins) = _midpoints(o.edges)
+_counts(o::FixedBins) = o.counts
+nobs(o::FixedBins) = sum(o.counts) + o.out
+function fit!(o::FixedBins, y, γ::Number)
+    idx = _binindex(o, y)
+    if 1 ≤ idx < length(o.edges)
+        @inbounds o.counts[idx] += 1
+    else
         o.out += 1
     end
 end
-function Base.merge!(o::FixedRangeBins, o2::FixedRangeBins, γ::Number) 
+
+function _binindex(o::FixedBins{:left}, y)
+    edges = o.edges
+    a = first(edges)
+    if y < a
+        return 0
+    elseif y == last(edges)
+        # right-most bin is a closed interval [a, b]
```

Comment:
Not that I've dug deep into histogram literature, but I've never seen the "end bin" as closed.  Shouldn't it be half-open?

Reply:
I've not studied histogram literature either. My samples that are restricted to [0,1] but I was worried that I might exclude some samples if all bins are half-closed (I've seen both values of 0 and 1, so this might actually happen). Moreover, MATLAB apparently avoids this issue by specifying the "end bin" as closed: https://se.mathworks.com/help/matlab/ref/histcounts.html

Reply:
It looks like R also closes the interval (with option to keep it half-open).  As far as I can tell, StatsBase leaves it open without the option to close it.  I think your use case is common, so I'm happy to leave it as is and we can always add the option later.